### ISSUE TITLE
User passwords visible for administrators

### ISSUE THREAD

**Issue by stweil** [kitodo/kitodo-production#626]
Administrators can enter the "Edit user" page for any registered user.
There they can see the password of individual users (with minimum efforts).

**Comment by henning-gerhardt** [comment]
Good catch but an administrator can take over user control (login in as this user) and can change user password if he wants or needed this to do (user password reset).

**Comment by stweil** [comment]
That's not the basic problem. The problem is that passwords are private information (and many people reuse them – which is bad), and that information must only be known to their owner.

**Comment by stweil** [comment]
Side remark: could somebody fix those bounty message additions? They must be marked as added by a robot (otherwise they must not appear under my name), or even better be written as a reply message (like those from CI). In addition the formatting is obviously broken.

**Comment by stweil** [comment]
There is another related problem: Up to now, I thought that only encrypted passwords were stored (see also issue #456), but this issue shows that the passwords are also stored in clear text. This is not state of the art and must be considered as a security issue.

**Comment by henning-gerhardt** [comment]
Please look into the database and see that the user passwords are stored encrypted! Even with LDAP usage user passwords are encrypted stored into database and LDAP system.

My guess: while loading user profile for displaying the user password is decrypted.

**Comment by stweil** [comment]
I fixed my comment and the issue title. The decryption must be irreversible, otherwise it is insecure.

**Comment by matthias-ronge** [comment]
Minimum requirements to stored passwords are that they are stored [hashed](https://en.wikipedia.org/wiki/Cryptographic_hash_function) by a hash function that is [cryptographically safe](https://en.wikipedia.org/wiki/Security_of_cryptographic_hash_functions). (This is different from (reversible) _encryption_.) In addition, each password must be stored with an _individual_ [salt](https://en.wikipedia.org/wiki/Salt_(cryptography)), making the generation of [rainbow tables](https://en.wikipedia.org/wiki/Rainbow_table) useless. Also, dumping and re-importing the user table must not break the hashes. (Both MySQL and Java handle illegal string content, but it cannot be restored from a MySQL dump, which is a text file.)

Since passwords cannot be decrypted, there is need for a recovery option. Typical solution is to e-mail the user a link with a one-time hash which allows them to reset their password. It should be clear that this password should only be usable once to change the password, and it should also expire after a short period (configurable, but I am talking about hours here). However, some desktop OSes (for example KDE) may call the link in the moment you click on it (probably to get the MIME type), to then decide to pass it to the browser. So the link must only be _usable_ once to change the password, but as long as it was not used to change the password, the page _must remain requestable_ until the time-out is reached. (What I want to say: The wrong solution is to make the page _requestable_ only once.)

**Comment by henning-gerhardt** [comment]
@matthias-ronge : good hint and explanation. This  should even go to issue #456.

**Comment by matthias-ronge** [comment]
I copy it there.
